src/psyclone/domain/lfric/psyir.py

```diff
@@ -0,0 +1,284 @@
+# Author R. W. Ford STFC Daresbury Lab
+
+'''This module generates LFRic-specific PSyIR classes from lists of
+definitions.
+
+'''
+# pylint: disable=unused-import
+# pylint: disable=exec-used
+from collections import namedtuple
+from psyclone.psyir.symbols import ContainerSymbol, DataSymbol, DeferredType, \
+    GlobalInterface, ScalarType, ArrayType
+from psyclone.psyir.nodes import Literal
+
+# Define LFRic module symbols.
+
+# The first Module namedtuple argument specifies the name of the
+# module and the second argument declares the name(s) of any symbols
+# declared by the module.
+
+Module = namedtuple('Module', ["name", "vars"])
+modules = [
+    Module("constants_mod", ["i_def", "r_def", "l_def"])]
+
+# Generate LFRic module symbols from definitions
+for module in modules:
+    MODULE_NAME = module.name
+    # Create the module (using a PSyIR ContainerSymbol)
+    exec("{0} = ContainerSymbol('{1}')\n".format(
+        MODULE_NAME.upper(), MODULE_NAME))
+    # Create the variables specified by the module (using PSyIR DataSymbols)
+    for module_var in module.vars:
+        exec("{0} = DataSymbol('{1}', DeferredType(), interface="
+             "GlobalInterface({2}))".format(
+                 module_var.upper(), module_var, MODULE_NAME.upper()))
+
+# Define generic LFRic scalar datatypes and symbols
+
+# The GenericScalar namedtuple has 3 properties: the first determines
+# the names of the resultant datatype and datasymbol classes, the
+# second specifies the intrinsic PSyIR type and the third specifies
+# the precision required by referencing symbols already declared
+# above.
+
+GenericScalar = namedtuple('GenericScalar', ["name", "intrinsic", "precision"])
+generic_scalar_datatypes = [
+    GenericScalar("lfric integer scalar", "integer", "i_def"),
+    GenericScalar("lfric real scalar", "real", "r_def"),
+    GenericScalar("lfric logical scalar", "boolean", "l_def")]
+
+# Generate generic LFRic scalar datatypes and symbols from definitions
+for info in generic_scalar_datatypes:
+    NAME = "".join(info.name.title().split())
+    INTRINSIC = info.intrinsic.upper()
+    PRECISION = info.precision
+    # Create the specific datatype
+    exec(
+        "class {0}DataType(ScalarType):\n"
+        "    def __init__(self, precision=None):\n"
+        "        if not precision:\n"
+        "            precision = {1}\n"
+        "        super({0}DataType, self).__init__(\n"
+        "            ScalarType.Intrinsic.{2}, precision)\n"
+        "".format(NAME, PRECISION.upper(), INTRINSIC))
+    # Create the specific symbol
+    exec(
+        "class {0}DataSymbol(DataSymbol):\n"
+        "    def __init__(self, name, precision=None, interface=None):\n"
+        "        super({0}DataSymbol, self).__init__(\n"
+        "            name, {0}DataType(precision=precision),\n"
+        "            interface=interface)\n"
+        "".format(NAME))
+
+
+# Define any LFRic-specific scalar literals
+class LfricDimension(Literal):
```

Comment:
pylint complains about abstract methods (`__str__` and `gen_code`) not being overridden.

Reply:
`gen_code()` is a generic `Node` problem. The original PSyIR for the PSy-layer code generation uses gen_code, but newer nodes do not (we use a visitor) so we chose to just ignore it. This will go away when we finally drop `gen_code()`.

`__str__` is similar in that many subclasses of `Node` have not implemented it. I'm not sure why it was made mandatory in the base class in the first place, although I think it was me that did that!

I don't think the answer is to fix it for this one case. We could raise yet another issue I suppose.

Reply:
Oh yes, you're right. I must have been in a particularly picky mood!